### ISSUE TITLE
Enforcing Unique IDs

### ISSUE THREAD

**Issue by Aryk** [tommikaikkonen/redux-orm#81]
Is this possible to do? Seems like redux-orm would be confused if there were multiple records with the same ID since it wouldn't know which one to fetch. Do you plan to add a check against this?

**Comment by tommikaikkonen** [comment]
Currently a creating a record with an id that equals another existing record would just overwrite it - that's not expected functionality. I agree that this should be added.

Ideally this will be added to the database layer [`Table.prototype.insert`](https://github.com/tommikaikkonen/redux-orm/blob/master/src/db/Table.js#L163). Currently that returns a new state and the created record. It should be extended to be able to return an error value as well, if it wasn't able to create a new record. Ability to return error values from the database layer isn't used by the ORM layer yet but when the new database API introduced in 0.9.0, it accommodates use of that pattern.

For now, you can patch this easily at the ORM layer before a full feature is added (might differ slightly for TS):

```javascript
class MyBaseModel extends Model {
  static create(props) {
    if (props.hasOwnProperty(this.idAttribute) && this.hasId(props[this.idAttribute])) {
      throw new Error(`Id with ${props[this.idAttribute]} already exists`);
    }
    return super.create(props);
  }
}
```

**Comment by Aryk** [comment]
yup, makes sense, just checking on this. That's what I added in my basemodel.

**Comment by carlesnunez** [comment]
I created the same object that u can extend to make and update if we currently have the record on our store or create if not:

```javascript
import { Model } from 'redux-orm';

export default class DopeModel extends Model {
  static create(props) {
    if (props.hasOwnProperty(this.idAttribute) && this.hasId(props[this.idAttribute])) {
      const model = this.withId(props.id);
      return model.update(props);
    }
    return super.create(props);
  }
}
```

**Comment by udayreddy** [comment]
can we have idAttribute with 2 columns or two props of the object ?

**Comment by haveyaseen** [comment]
@udayreddy This is currently not possible. And I'm not planning this myself because there's already more than enough on the roadmap. But would happily merge a PR if tested, consistent with current memoization and non-breaking. I guess a reasonable API would be `Book.idAttribute = ["title", "edition"]`.

The obvious easy workaround would be to store them in the same field instead (`` book.id = `${book.title}-${book.edition}` ``). Note that that's likely to cause bugs as you need to do this manually each time you perform a write operation (`book.id = someId`, `book.set("id", someId)`, `book.update({ id: someId })`, `Book.create`). So if you can, I would suggest to add a surrogate key to your data source for this purpose.

And for anyone passing by, the `Model.upsert` solution by @carlesnunez has been added in v0.9.1.

**Comment by carlesnunez** [comment]
Thank you for the kudos @haveyaseen 🤟🏻😊
